libs/langchain_v1/langchain/agents/middleware/tool_call_limit.py

```diff
@@ -277,8 +281,76 @@ def after_model(self, state: ToolCallLimitState, runtime: Runtime) -> dict[str,
         run_counts = state.get("run_tool_call_count", {}).copy()
 
         # Increment counts for this key
-        thread_counts[count_key] = thread_counts.get(count_key, 0) + tool_call_count
-        run_counts[count_key] = run_counts.get(count_key, 0) + tool_call_count
+        new_thread_count = thread_counts.get(count_key, 0) + tool_call_count
+        new_run_count = run_counts.get(count_key, 0) + tool_call_count
+
+        thread_counts[count_key] = new_thread_count
+        run_counts[count_key] = new_run_count
+
+        # Check if any limits are exceeded after incrementing
+        thread_limit_exceeded = (
+            self.thread_limit is not None and new_thread_count > self.thread_limit
+        )
+        run_limit_exceeded = self.run_limit is not None and new_run_count > self.run_limit
+
+        if thread_limit_exceeded or run_limit_exceeded:
+            limit_message = _build_tool_limit_exceeded_message(
+                new_thread_count, new_run_count, self.thread_limit, self.run_limit, self.tool_name
+            )
+
+            if self.exit_behavior == "error":
+                raise ToolCallLimitExceededError(
+                    thread_count=new_thread_count,
+                    run_count=new_run_count,
+                    thread_limit=self.thread_limit,
+                    run_limit=self.run_limit,
+                    tool_name=self.tool_name,
+                )
+
+            # For both "continue" and "end", inject artificial error ToolMessages
+            # for tool calls that exceeded their limits
+            artificial_messages: list[ToolMessage | AIMessage] = []
+            for tool_call in last_ai_message.tool_calls:
+                # Only inject errors for tool calls that match our filter
+                if self.tool_name is None or tool_call["name"] == self.tool_name:
+                    artificial_messages.extend(
+                        [
+                            ToolMessage(
+                                content=limit_message,
+                                tool_call_id=tool_call["id"],
+                                name=tool_call.get("name"),
+                                status="error",
+                            )
+                        ]
+                    )
+
+            if self.exit_behavior == "end":
+                # For "end" behavior, only support single tool call scenarios
+                if len(last_ai_message.tool_calls) > 1:
```

Comment:
If all tool calls are the same tool, we can just apply the same tool message to all of them